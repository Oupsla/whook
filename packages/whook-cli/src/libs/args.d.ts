import { WhookCommandDefinitionArguments } from '../services/promptArgs';
import { WhookCommandArgs } from '../services/args';
export declare function readArgs(
  schema: WhookCommandDefinitionArguments,
  args: WhookCommandArgs,
): WhookCommandArgs;
