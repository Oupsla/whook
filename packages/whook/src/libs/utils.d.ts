export declare function noop(...args: any[]): void;
export declare function identity<T>(id: T): T;
export declare function compose(...fns: Function[]): Function;
export declare function pipe(...fns: Function[]): Function;
