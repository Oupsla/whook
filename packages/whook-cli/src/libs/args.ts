import Ajv from 'ajv';
import YError from 'yerror';
import { WhookCommandDefinitionArguments } from '../services/promptArgs';
import { WhookCommandArgs } from '../services/args';

// TODO: Add ajv human readable error builder
export function readArgs(
  schema: WhookCommandDefinitionArguments,
  args: WhookCommandArgs,
): WhookCommandArgs {
  const ajv = new Ajv({
    coerceTypes: true,
    useDefaults: true,
    strictDefaults: true,
    strictKeywords: true,
  });
  const validator = ajv.compile(schema);

  const {
    _: [, ...listedArgs],
    ...namedArgs
  } = args;
  const cleanedArgs = {
    ...namedArgs,
    ...(listedArgs.length ? { _: listedArgs } : {}),
  } as WhookCommandArgs;

  validator(cleanedArgs);

  if ((validator.errors || []).length) {
    throw new YError('E_BAD_ARGS', validator.errors);
  }

  return cleanedArgs;
}
