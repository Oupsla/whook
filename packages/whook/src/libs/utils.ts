// @ts-ignore: noop do not operate ¯\_(ツ)_/¯
export function noop(...args: any[]) {}

export function identity<T>(id: T): T {
  return id;
}

export function compose(...fns: Function[]): Function {
  return fns.reduce((f, g) => (...args: Array<any>) => f(g(...args)));
}

export function pipe(...fns: Function[]): Function {
  return compose.apply(compose, fns.reverse());
}
