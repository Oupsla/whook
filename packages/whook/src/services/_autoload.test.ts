import initAutoload from './_autoload';
import { service } from 'knifecycle';
import YError from 'yerror';
import { identity } from '../libs/utils';

describe('$autoload', () => {
  const log = jest.fn();
  const $injector = jest.fn();
  const require = jest.fn();
  const resolve = jest.fn();

  beforeEach(() => {
    log.mockReset();
    $injector.mockReset();
    require.mockReset();
    resolve.mockReset();
  });

  describe('should work', () => {
    it('for configs', async () => {
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/services/CONFIGS.js',
      );
      require.mockReturnValueOnce({
        default: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('CONFIGS');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for a config constant', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('CONFIG');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for API', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/services/API.js',
      );
      require.mockImplementationOnce(() => ({
        default: service(async () => ({ info: {} }), 'API'),
      }));

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('API');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for SERVICE_NAME_MAP', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js',
      );
      require.mockImplementationOnce(() => ({
        default: service(async () => ({ info: {} }), 'SERVICE_NAME_MAP'),
      }));

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('SERVICE_NAME_MAP');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for handlers hash', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: {
          openapi: '3.0.2',
          info: {
            version: '1.0.0',
            title: 'Sample OpenAPI',
            description: 'A sample OpenAPI file for testing purpose.',
          },
          paths: {
            '/ping': {
              get: {
                operationId: 'getPing',
                summary: "Checks API's availability.",
                responses: {
                  '200': {
                    description: 'Pong',
                  },
                },
              },
            },
          },
        },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      require.mockReturnValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('HANDLERS');

      expect({
        result,
        HANDLERS: await result.initializer({ getPing: () => {} }),
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      require.mockReturnValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for name mapped handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPingMock.js',
      );
      require.mockReturnValueOnce({
        default: service(
          async () => async () => ({ status: 200 }),
          'getPingMock',
        ),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {
          getPing: 'getPingMock',
        },
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for name mapped handlers with dynamic SERVICE_NAME_MAP', async () => {
      $injector.mockResolvedValueOnce({
        SERVICE_NAME_MAP: {
          getPing: 'getPingMock',
        },
      });
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPingMock.js',
      );
      require.mockReturnValueOnce({
        default: service(
          async () => async () => ({ status: 200 }),
          'getPingMock',
        ),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for path mapped handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      require.mockReturnValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {
          getPing: '/home/whoami/my-other-project/src/handlers/getPing.js',
        },
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for handlers in sub plugins', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockImplementationOnce(() => {
        throw new YError('E_BAD_MODULE');
      });
      resolve.mockImplementationOnce(
        () => '/var/lib/node/node_modules/@whook/cli/dist/handlers/getPing.js',
      );
      require.mockReturnValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [
          '/var/lib/node/node_modules/@whook/cli/dist',
          '/var/lib/node/node_modules/@whook/lol/dist',
        ],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('with no wrappers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      require.mockImplementationOnce(() => ({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      }));

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('with empty wrappers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        WRAPPERS: [],
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      require.mockReturnValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPingWrapped');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for wrapped handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        WRAPPERS: [identity],
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      require.mockReturnValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });
      const result = await $autoload('getPingWrapped');

      expect({
        result,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
        requireCalls: require.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });
  });

  describe('should fail', () => {
    it('with unexisting handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockImplementationOnce(() => {
        throw new YError('E_NO_MODULE');
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        require: (require as unknown) as any,
        resolve: (resolve as unknown) as any,
      });

      try {
        await $autoload('getPing');
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
          injectorCalls: $injector.mock.calls,
          requireCalls: require.mock.calls,
          resolveCalls: resolve.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
