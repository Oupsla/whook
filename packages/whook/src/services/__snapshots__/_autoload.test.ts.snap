// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$autoload should fail with unexisting handlers 1`] = `
Object {
  "errorCode": "E_UNMATCHED_DEPENDENCY",
  "errorParams": Array [
    "getPing",
  ],
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸš« - Service \\"getPing\\" not found in: /home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "requireCalls": Array [],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
}
`;

exports[`$autoload should work for API 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"API\\" found in: /home/whoami/my-whook-project/src/services/API.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading API initializer from /home/whoami/my-whook-project/src/services/API.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/services/API.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/services/API",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "API",
    "path": "/home/whoami/my-whook-project/src/services/API.js",
  },
}
`;

exports[`$autoload should work for SERVICE_NAME_MAP 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"SERVICE_NAME_MAP\\" found in: /home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading SERVICE_NAME_MAP initializer from /home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "SERVICE_NAME_MAP",
    "path": "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js",
  },
}
`;

exports[`$autoload should work for a config constant 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
  ],
  "requireCalls": Array [],
  "resolveCalls": Array [],
  "result": Object {
    "initializer": [Function],
    "name": "CONFIG",
    "path": "internal://CONFIG",
  },
}
`;

exports[`$autoload should work for configs 1`] = `
Object {
  "injectorCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"CONFIGS\\" found in: /home/whoami/my-whook-project/src/services/CONFIGS.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading CONFIGS initializer from /home/whoami/my-whook-project/src/services/CONFIGS.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/services/CONFIGS.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/services/CONFIGS",
    ],
  ],
  "result": Object {
    "initializer": Object {
      "CONFIG": Object {
        "testConfig": "test",
      },
    },
    "name": "CONFIGS",
    "path": "/home/whoami/my-whook-project/src/services/CONFIGS.js",
  },
}
`;

exports[`$autoload should work for handlers 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPing\\" found in: /home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPing initializer from /home/whoami/my-whook-project/src/handlers/getPing.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work for handlers hash 1`] = `
Object {
  "HANDLERS": Object {
    "getPing": [Function],
  },
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
    Array [
      Array [
        "API",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
  ],
  "requireCalls": Array [],
  "resolveCalls": Array [],
  "result": Object {
    "initializer": [Function],
    "name": "HANDLERS",
    "path": "internal://HANDLERS",
  },
}
`;

exports[`$autoload should work for handlers in sub plugins 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸš« - Service \\"getPing\\" not found in: /home/whoami/my-whook-project/src/handlers/getPing",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPing\\" found in: /var/lib/node/node_modules/@whook/cli/dist/handlers/getPing.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPing initializer from /var/lib/node/node_modules/@whook/cli/dist/handlers/getPing.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/var/lib/node/node_modules/@whook/cli/dist/handlers/getPing.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/cli/dist/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPing",
    "path": "/var/lib/node/node_modules/@whook/cli/dist/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work for name mapped handlers 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ“– - Using SERVICE_NAME_MAP to route \\"getPing\\" to \\"getPingMock\\".",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPingMock\\" found in: /home/whoami/my-whook-project/src/handlers/getPingMock.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPing initializer via getPingMock resolution from /home/whoami/my-whook-project/src/handlers/getPingMock.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPingMock",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPingMock",
    "path": "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
  },
}
`;

exports[`$autoload should work for name mapped handlers with dynamic SERVICE_NAME_MAP 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "SERVICE_NAME_MAP",
      ],
    ],
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ“– - Using SERVICE_NAME_MAP to route \\"getPing\\" to \\"getPingMock\\".",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPingMock\\" found in: /home/whoami/my-whook-project/src/handlers/getPingMock.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPing initializer via getPingMock resolution from /home/whoami/my-whook-project/src/handlers/getPingMock.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPingMock",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPingMock",
    "path": "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
  },
}
`;

exports[`$autoload should work for path mapped handlers 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ“– - Using INITIALIZER_PATH_MAP to resolve the \\"getPing\\" module path.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPing\\" found in: /home/whoami/my-other-project/src/handlers/getPing.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPing initializer from /home/whoami/my-other-project/src/handlers/getPing.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-other-project/src/handlers/getPing.js",
    ],
  ],
  "resolveCalls": Array [],
  "result": Object {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-other-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work for wrapped handlers 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
    Array [
      Array [
        "WRAPPERS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPingWrapped\\" found in: /home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPingWrapped initializer from /home/whoami/my-whook-project/src/handlers/getPing.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPingWrapped",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work with empty wrappers 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
    Array [
      Array [
        "WRAPPERS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPingWrapped\\" found in: /home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPingWrapped initializer from /home/whoami/my-whook-project/src/handlers/getPing.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPingWrapped",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work with no wrappers 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "CONFIGS",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Service \\"getPing\\" found in: /home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
    Array [
      "debug",
      "ðŸ’¿ - Loading getPing initializer from /home/whoami/my-whook-project/src/handlers/getPing.js.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;
